### DISCUSSION TITLE
new outline installation, db migration keeps failing

### DISCUSSION THREAD

**Discussion by leonsteinhaeuser** [outline/outline#4143] [Self Hosting] [answered]
Currently, when I try to deploy outline, I encounter the error below during the database migration.

**To Reproduce**
1. use the configuration from below
2. `yarn sequelize db:migrate`
3. See error

**Expected behavior**
I expect the migration to complete.

**Outline (please complete the following information):**
- Install: self hosted
- Version: v0.66.0

Config (env):
```bash
/opt/outline $ printenv
COLLABORATION_URL=
DATABASE_URL=postgres://XXX@outline-zgr2o94pxa-postgresql-ha-pgpool.outline-zgr2o94pxa.svc:5432/outline?sslmode=disable&password=XXX
NODE_VERSION=16.14.2
HOSTNAME=outline-zgr2o94pxa-5744dc6b97-8vdtt
SECRET_KEY=XXX
URL=http://outline.example.com
DEBUG=http
YARN_VERSION=1.22.18
SHLVL=1
AWS_S3_UPLOAD_BUCKET_URL=s3.example.com
PORT=3000
HOME=/home/nodejs
UTILS_SECRET=XXX
TEAM_LOGO=
AWS_S3_FORCE_PATH_STYLE=1e+07
ENABLE_UPDATES=false
AWS_S3_ACL=private
AWS_S3_UPLOAD_BUCKET_NAME=outline
WEB_CONCURRENCY=1
TERM=xterm
AWS_S3_UPLOAD_MAX_SIZE=1e+07
DEFAULT_LANGUAGE=eu_US
GOOGLE_ANALYTICS_ID=
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
MAXIMUM_IMPORT_SIZE=5.12e+06
AWS_REGION=us-east-1
AWS_ACCESS_KEY_ID=XXX
AWS_SECRET_ACCESS_KEY=XXX
SENTRY_DSN=
ALLOWED_DOMAINS=
PWD=/opt/outline
DATABASE_CONNECTION_POOL_MIN=2
FORCE_HTTPS=false
REDIS_URL=redis://outline-zgr2o94pxa-redis-master.outline-zgr2o94pxa.svc:6379
DATABASE_CONNECTION_POOL_MAX=10
NODE_ENV=production-ssl-disabled
PGSSLMODE=disable
````

Error:
```bash
/opt/outline $ yarn sequelize db:migrate --env production-ssl-disabled
yarn run v1.22.18
$ /opt/outline/node_modules/.bin/sequelize db:migrate --env production-ssl-disabled

Sequelize CLI [Node: 16.14.2, CLI: 6.4.1, ORM: 6.20.1]

Loaded configuration file "server/config/database.json".
Using environment "production-ssl-disabled".
node:internal/crypto/keys:570
    throw new ERR_INVALID_ARG_TYPE(
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "key" argument must be of type string or an instance of ArrayBuffer, Buffer, TypedArray, DataView, KeyObject, or CryptoKey. Received null
    at new NodeError (node:internal/errors:371:5)
    at prepareSecretKey (node:internal/crypto/keys:570:11)
    at new Hmac (node:internal/crypto/hash:132:9)
    at Object.createHmac (node:crypto:162:10)
    at createHMAC (/opt/outline/node_modules/pg/lib/sasl.js:133:17)
    at Hi (/opt/outline/node_modules/pg/lib/sasl.js:137:13)
    at Object.continueSession (/opt/outline/node_modules/pg/lib/sasl.js:32:24)
    at Client._handleAuthSASLContinue (/opt/outline/node_modules/pg/lib/client.js:257:10)
    at Connection.emit (node:events:526:28)
    at /opt/outline/node_modules/pg/lib/connection.js:115:12
    at Parser.parse (/opt/outline/node_modules/pg-protocol/dist/parser.js:40:17)
    at Socket.<anonymous> (/opt/outline/node_modules/pg-protocol/dist/index.js:10:42)
    at Socket.emit (node:events:526:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10) {
  code: 'ERR_INVALID_ARG_TYPE'
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by vgwidt** [comment]
```
TypeError [ERR_INVALID_ARG_TYPE]: The "key" argument must be of type string or an instance of ArrayBuffer, Buffer, TypedArray, DataView, KeyObject, or CryptoKey. Received null
    at new NodeError (node:internal/errors:371:5)
    at prepareSecretKey (node:internal/crypto/keys:570:11)
```

You need to make sure necessary environment variables for Outline are set such as SECRET_KEY.  You can see the sample here: https://github.com/outline/outline/blob/main/.env.sample

**Comment by leonsteinhaeuser** [comment]
@vgwidt thanks for your reply. I filtered the output a bit to remove the K8s specific environment variables. It seems that I also removed a few outline specific variables. In the "original" environment secret_key and utils_secret are set.
I have updated the list above.

The definition for the variables can be found here:
- https://github.com/programmfabrik/outline-helm/blob/main/charts/outline/templates/secret.yaml
- https://github.com/programmfabrik/outline-helm/blob/main/charts/outline/templates/configmap.yaml

**Comment by leonsteinhaeuser** [comment, marked as answer]
@tommoor I saw that you have turned my issue into a discussion. Can you please provide a solution to the above error.

**Reply by tommoor** [reply]
Most likely you need to set `PGSSLMODE=disable` in env

**Reply by leonsteinhaeuser** [reply]
I added `PGSSLMODE=disable` but it still fails with the same message as above.

**Reply by tommoor** [reply]
Why do you have `password=` appended to the postgres url instead of using the standard `username:password` formatting for this?

**Reply by leonsteinhaeuser** [reply]
Because I have special characters in my password that may not be valid in the username:password@url format.

**Reply by leonsteinhaeuser** [reply]
I have changed the password in my test environment and it seems that this is the problem. So the parameter password=outline is not read.

**Reply by tommoor** [reply]
I'd suggest you try changing this next

**Reply by leonsteinhaeuser** [reply]
Is it possible to support this?

**Reply by tommoor** [reply]
Short answer, no. Long answer - you'd have to change the underlying implementation of connection string parsing in one of the most popular node libraries: https://github.com/brianc/node-postgres/tree/master/packages/pg-connection-string

**Reply by leonsteinhaeuser** [reply]
Well, that's unfortunate, but I think I can find a way to provide the password in the "correct" format.

Many thanks for your help.
